File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        ArrayList<String> list = new ArrayList<>();
        list.add("+");
        list.add("-");
        list.add("*");
        list.add("/");
        int x;
        int y;
        int exp;
        // char[] words = tokens.toCharArray();
        for (String i : tokens) {
            if (list.contains(i) && stack.size() > 1) {
                x = stack.pop();
                y = stack.pop();
                switch(i) {
                    case "+":
                        exp = y + x;
                        break;
                    case "-":
                        exp = y - x;
                        break;
                    case "*":
                        exp = y * x;
                        break;
                    case "/":
                        exp = y / x;
                        break;
                    default:
                        exp = 0;
                        break;
                }
                stack.push(exp);
            }
            else {
                stack.push(Integer.parseInt(i));
            }
        }
        return stack.pop();
    }
}

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
class Solution {
static constexpr int division(int a, int b) {
    return a / b;
}

static constexpr int multiplication(int a, int b) {
    return a * b;
}

static constexpr int addition(int a, int b) {
    return a + b;
}

static constexpr int soustraction(int a, int b) {
    return a - b;
}

public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stack;
        unordered_map<string, std::function<int(int, int)>> map;
        map["+"] = addition;
        map["-"] = soustraction;
        map["/"] = division;
        map["*"] = multiplication;

        for (string s : tokens) {
            if (map.find(s) == map.end()) {
                stack.push(stoi(s));
            } else {
                auto first = stack.top();
                stack.pop();
                auto second = stack.top();
                stack.pop();
                stack.push(map[s](second, first));
            }
        }

        return stack.top();
    }
};